lib/os_io.c

```diff
@@ -37,6 +37,130 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_poll
+#define ___fdset_state_size(ps)     ps->os.fdset.size
+#define ___fdset_state_readfds(ps)  ps->os.fdset.readfds
+#define ___fdset_state_writefds(ps) ps->os.fdset.writefds
+
+static int ___fdset_heap_overflow;
+
+static int ___fdset_state_init (___processor_state ps)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+
+  readfds  = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!readfds)
+    goto error;
+
+  writefds = ___ALLOC_MEM (MAX_CONDVARS/8);
+  if (!writefds)
+    goto error;
+
+  memset (readfds, 0, MAX_CONDVARS/8);
+  memset (writefds, 0, MAX_CONDVARS/8);
+  ___fdset_state_readfds (ps) = readfds;
+  ___fdset_state_writefds (ps) = writefds;
+  ___fdset_state_size (ps) = MAX_CONDVARS;
+
+  return 0;
+
+ error:
+  ___free_mem (readfds);
+  ___free_mem (writefds);
+  return 1;
+}
+
+static int ___fdset_realloc (___processor_state ps, int fd)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  int oldsize = ___fdset_state_size (ps);
+  int newsize = oldsize;
+
+  while (newsize <= fd)
+    newsize = newsize * 2;
+
+  if (oldsize == newsize) /* we never shrink fdsets */
```

Comment:
Comment is strange... the logic of the while is what makes the size never shrink, not the test ```oldsize == newsize``` (comment should be "optimize for when size doesn't change").